contracts/exchangeIssuance/FlashMintWrapped.sol

```diff
@@ -17,220 +17,905 @@
+   * @param _amount               Amount that will be bought
+   * @param _maxAmountIn          Maximum aount of input token that can be spent
+   * @param _swapDexData          DEXAdapter.SwapData with path, fees, etc. for inputToken -> outputToken swap
+   *
+   * @return Amount of spent _inputToken
+   */
+  function _swapToExact(
+    IERC20 _inputToken,
+    IERC20 _outputToken,
+    uint256 _amount,
+    uint256 _maxAmountIn,
+    DEXAdapter.SwapData calldata _swapDexData
+  )
+    internal
+    isValidPath(_swapDexData.path, address(_inputToken), address(_outputToken))
     returns (uint256)
-    {
-        return 0;
+  {
+    // safe approves are done right in the dexAdapter library
+    return dexAdapter.swapTokensForExactTokens(_amount, _maxAmountIn, _swapDexData);
+  }
+
+  /**
+   * Swaps exact _amount of _inputToken to _outputToken through _swapDexData
+   *
+   * @param _inputToken           Input token that will be sold
+   * @param _outputToken          Output token that will be bought
+   * @param _amount               Amount that will be sold
+   * @param _swapDexData          DEXAdapter.SwapData with path, fees, etc. for inputToken -> outputToken swap
+   *
+   * @return amount of received _outputToken
+   */
+  function _swapFromExact(
+    IERC20 _inputToken,
+    IERC20 _outputToken,
+    uint256 _amount,
+    DEXAdapter.SwapData calldata _swapDexData
+  )
+    internal
+    isValidPath(_swapDexData.path, address(_inputToken), address(_outputToken))
+    returns (uint256)
+  {
+    // safe approves are done right in the dexAdapter library
+    return
+      dexAdapter.swapExactTokensForTokens(
+        _amount,
+        // _minAmountOut is 0 here since we don't know what to check against because for wrapped components
+        // we only have the required amounts for the wrapped component, but not for the underlying we swap to here
+        // This is covered indirectly in later checks though
+        // e.g. directly through the issue call (not enough _outputToken -> wrappedComponent -> issue will fail)
+        0,
+        _swapDexData
+      );
+  }
+
+  /**
+   * Wraps _wrapAmount of _underlyingToken to _wrappedComponent component
+   *
+   * @param _wrappedComponent           Address of wrapped component (e.g. cDAI)
+   * @param _wrapAmount                 amount of _underlyingToken to wrap
+   * @param _underlyingToken            underlying (unwrapped) token to wrap from (e.g. DAI)
+   * @param _wrapData                   ComponentWrapData that contains the integration (adapter) name and optional bytes data
+   */
+  function _wrapComponent(
+    address _wrappedComponent,
+    uint256 _wrapAmount,
+    address _underlyingToken,
+    ComponentWrapData calldata _wrapData
+  ) internal {
+    // 1. get the wrap adapter directly from the integration registry
+
+    // Note we could get the address of the adapter directly in the params instead of the integration name
```

Comment:
See ExchangeIssuanceZeroEx where we're forced to pass in the bytes calldata for the initial swap function. Truth to be told, the reason I want to directly avoid using the wrap adapters is to save a lot of gas on all the contract calls. In saying that, we could actually have a view function in this contract which uses the wrap adapters to generate all the swap data params and then have the client pass the raw bytes passed into the execute function.

However, we'll still need to take in raw bytes for the initial exchange (such as with 0x protocol).

Reply:
I think you are right about saving gas, I just ran some rough numbers and getting the adapter and call data costs about 15k gas.

The difference with ExchangeIssuanceZeroEx is that we _know_ and _define_ the swapTarget on which we invoke.
I am not really comfortable with calling out to an unknown contract for security reasons.

_But_ there is a simple solution to that: 
For Compound and Yearn the wrap / unwrap calls are invoked directly on the wrappedToken address. 
For Aave there is one lendingPool contract which all wrapping / unwrapping methods go to.
So:
`if wrapCallTarget == SetToken component ||  allowedWrapCallTargets[wrapCallTarget]`
should do the trick.

We add a mapping allowedWrapCallTargets that can be edited by the contract owner.
SetToken components are indirectly allowListed, someone would have to be able to add components to the SetToken

No drawbacks on security + gas cost saved... What do you think?

Reply:
@FlattestWhite

Reply:
Yeah having an allowlist of addresses that we're allowed to call was what I was thinking as well. An SLOAD is cheaper than a contract call after all. We'll have to make this contract pausable, ownable so that we can add and remove different allowlisted contracts.

Reply:
yea I have it Withdrawable (which includes Ownable) 

hmm why Pausable @FlattestWhite? No other Set / IndexCoop contract is Pausable AFAIK

Reply:
As mentioned in my review only verifying the callTarget and using user specified calldata seems dangerous to me. Especially if the user will be able to encode `approve` or `transfer` calls in this calldata.

Reply:
The original implementation, calling the adapters directly in the issuance / redemption tx seems more robust and secure to me. Allowing the user to meddle with the calldata of chain opens various issues outlined below.

Reply:
@ckoopmann you do have a point. 

I think most of the things now discussed in this PR come down to how important it is to protect the potential left over dust in this contract... There shouldn't be any other funds in this contract? 

Seems like if we want to fight to keep this dust in this contract we need additional checks and we probably end up with the same gas cost as using the wrap adapters directly 

If we don't add those checks we put security at risk because of potential unknowns

Reply:
@FlattestWhite what's your opinion on this after @ckoopmann inputs?

Reply:
Resolving because discussed in other places